### ISSUE TITLE
Excanvas after-init event

### ISSUE THREAD

**Issue by GoogleCodeExporter** [josejaner/flot#133]
```
Excanvas initializes all canvas elements on the document.ready event, but
as it doesn't fire any event after it is done, it's annoying to draw things
in document.ready / $().

Adding this on line 126:
      if (window.jQuery) {
        window.jQuery().trigger('canvasready');
      }

and at the end of the file:
else {
  if (window.jQuery) {
    window.jQuery(function() { window.jQuery().trigger('canvasready'); });
  }
}

solves this problem with a custom event - you can start drawing whenever
canvasready triggers. You *need* this event to reliably draw things on
document.ready (setTimeout doesn't count).

I'm wondering if there are others who are interested in this functionality
as well. It would be handy to have it included in flot as I use it but also
draw directly on canvases, it saves me another thing to maintain and others
some headache in the future.
```

Original issue reported on code.google.com by `Boo...@gmail.com` on 14 Mar 2009 at 10:38

**Comment by GoogleCodeExporter** [comment]
```
I have a custom tooltip function that I bind in document.ready: 
$('#divid').bind('plothover', function () { ... }) that 
works fine on my line graphs in Firefox and Safari, but not in IE using 
excanvas.

As in the sample code, my function just appends a div to the body at the 
appropriate position containing the 
tooltip text (and removes it again when 'item' is false).

In IE, the first point hover works (I get the colored circle around the point), 
but the tooltip never appears and no 
additional points will subsequently highlight. 'hoverable' is set to true on 
the grid.

Is this perhaps a related issue?
```

Original comment by `rich.pol...@gmail.com` on 17 Mar 2009 at 9:18

**Comment by GoogleCodeExporter** [comment]
```
Hmm, I don't think so - we're using plothover as well to display tooltips, it 
works
on IE just as well as any other browser. But the method we use is different - we
don't insert divs dynamically, just show + position a pre-created tooltip div.
```

Original comment by `Boo...@gmail.com` on 18 Mar 2009 at 12:23

**Comment by GoogleCodeExporter** [comment]
```
I didn't understood this report, but I've reread it a couple of times, and as 
far as
I can tell, you're reporting a problem with excanvas. The right place to do 
that is
here (or on that project's mailing list):

  http://code.google.com/p/explorercanvas/issues/list

I don't maintain excanvas, I just bundle it with Flot to save people the hassle 
of
finding it themselves.

Note that when you call $.plot, Flot inserts a canvas element dynamically and 
inits
it manually on IE with

 window.G_vmlCanvasManager.initElement(c)

You could probably do the same. I'll close this now, if I misunderstood you
completely, feel free to reopen.


Rich Pollock: it sounds to me like you got a bug in your code. From my 
experience,
there are plenty of things that'll work fine in Firefox (and other browsers) but
break in IE.
```

Original comment by `olau%iol...@gtempaccount.com` on 21 Mar 2009 at 6:50
- Changed state: **Invalid**

**Comment by GoogleCodeExporter** [comment]
```
This issue gave me a bit of an headache today - race conditions are bad. Okay, 
so 
what happened is this:

- Excanvas register a 'onreadystatechange' handler. Only after its init is 
completed, G_vmlCanvasManager is available.
- I used the jquery document.ready event to start drawing things with Flot. It 
works 
when it runs after the Excanvas init, but that is not guaranteed by anything. 
In IE7, 
somehow, it works out alright most of the time. in IE8 standards mode, it 
doesn't.

I've added a $().trigger('canvasready') line to my local copy of excanvas, but 
it 
would be easier if it was integrated in flot, in its copy of excanvas. That's 
what I 
meant to say, as $(drawWithFlot) doesn't work on IE8, you need to use 
$().bind('canvasready', drawWithFlot).

```

Original comment by `Boo...@gmail.com` on 10 Apr 2009 at 11:30

**Comment by GoogleCodeExporter** [comment]
```
Issue 146 has been merged into this issue.
```

Original comment by `olau%iol...@gtempaccount.com` on 15 Apr 2009 at 5:31

**Comment by GoogleCodeExporter** [comment]
```
OK, so now I understand your problem (great analysis!). But I'm still not the
excanvas maintainer, and it clearly needs to be addressed at the excanvas level.

I think the idea of adding an extra event should be a last resort, I posted 
another
idea just now on the excanvas mailing list:

http://groups.google.com/group/google-excanvas/browse_thread/thread/7e2065266af1
cedd

Comments, tests would be appreciated, let's see if we can get them to fix it.
```

Original comment by `olau%iol...@gtempaccount.com` on 15 Apr 2009 at 6:03
- Changed state: **New**

**Comment by GoogleCodeExporter** [comment]
```
I hope the excanvas guys can provide a fix in their code, but a flot "branch" 
with
only one patch applied wouldn't hurt either. Either way, the weird race 
condition is
gone and developers are happy again ^^

Oh btw, there is a newer release available of excanvas than the one in flot 
svn. r3
contains mostly IE8 fixes, I've been using it without any problem so far on
http://www.worldoflogs.com.
```

Original comment by `Boo...@gmail.com` on 16 Apr 2009 at 4:48

**Comment by GoogleCodeExporter** [comment]
```
Seems Steve Clay had a good idea, we just call init_ in excanvas ourselves. I've
added the call to the latest revision, if you care, please give it a try:

http://flot.googlecode.com/svn/trunk/jquery.flot.js

I updated the bundled excanvas to the version 3 release, although I think I 
already
got the important changes with the latest SVN import.
```

Original comment by `olau%iol...@gtempaccount.com` on 28 Apr 2009 at 5:59

**Comment by GoogleCodeExporter** [comment]
```
Someone in issue 140 tested it, seems to work. So I'll close this bug now, 
thanks for
the analysis! Please comment again if you find it doesn't work.
```

Original comment by `olau%iol...@gtempaccount.com` on 7 May 2009 at 5:37
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
Hi Booink,
  Is the solution now integrated in flot0.7 version? I still get this problem on IE8 where the chart does not get displayed sometimes. Do we have to add the trigger canvas ready to our local copy of excanvas, if so where do we have to add it?? 

Thanks.
```

Original comment by `banugop...@gmail.com` on 13 Sep 2011 at 9:39

**Comment by GoogleCodeExporter** [comment]
```
I am also getting this exception.  Anything I can do?
```

Original comment by `stavaro...@gmail.com` on 8 Dec 2011 at 11:53
